### ISSUE TITLE
add command-line switch to scsynth to listen on a specific interface

### ISSUE THREAD

**Issue by jweather** [supercollider/supercollider#1266]
Suggestion:

`-L <IP to listen on> (default 0.0.0.0)`

This will be useful for Sonic Pi to avoid hassles with Windows firewall rules on installation by only listening on localhost, and may be useful to others for security purposes, or to use the same port on multiple interfaces.

**Comment by samaaron** [comment]
+1 here. 

Currently Sonic Pi has to ship with a modified version of SuperCollider so that schools can install it without triggering their firewalls. Clearly it would be much better to be using an official release rather than one we've hand rolled ourselves.

**Comment by telephon** [comment]
We need to make that properly switchable, because you don't want to disallow network music by default. I suppose it should be a server option.

**Comment by samaaron** [comment]
Agreed :)

**Comment by samaaron** [comment]
@jweather would a command line option work for us?

**Comment by jweather** [comment]
Yes, that would be ideal

**Comment by samaaron** [comment]
Would it be at all possible consider associating this to a concrete milestone such as `3.7.2`?

A feature such as this should be backwards compatible and therefore appropriate for a dot release.

**Comment by jweather** [comment]
This is the patch necessary to bind to localhost interface instead of all interfaces.  I do not have a patch dealing with the command-line switch logic yet:

````
diff --git a/server/scsynth/SC_ComPort.cpp b/server/scsynth/SC_ComPort.cpp
index 0e2b981..131e9e3 100644
--- a/server/scsynth/SC_ComPort.cpp
+++ b/server/scsynth/SC_ComPort.cpp
@@ -256,7 +256,7 @@ public:
 		BOOST_AUTO(protocol, ip::udp::v4());
 		udpSocket.open(protocol);
 
-		udpSocket.bind(ip::udp::endpoint(protocol, inPortNum));
+		udpSocket.bind(ip::udp::endpoint(boost::asio::ip::address::from_string(std::string("127.0.0.1")), inPortNum));
 
 		boost::asio::socket_base::send_buffer_size option(65536);
 		udpSocket.set_option(option);
````

**Comment by telephon** [comment]
So will the server accept messages from other machines on the network? It looks as if it was now locked to loopback.

**Comment by samaaron** [comment]
@telephon I think this patch does lock to loopback. This is what we need for Sonic Pi to stop `scsynth` from triggering the firewall in schools. The Windows version of Sonic Pi actually ships with a modified binary which contains this patch.

However, the ideal solution would be to have an additional command line flag which would allow you to start `scsynth` locked to loopback. Obviously this wouldn't be the default (to maintain consistency) :-)

**Comment by telephon** [comment]
Yes, it indeed wouldn't be a good idea to lock to loopback in general, as this would mean an end of most existing network music ensembles :)

You could just add a commandline option like `receiveOnlyFrom = nil;` when set to a non-nil value it would specify an address that is a unique controller.

**Comment by samaaron** [comment]
No intention here of inhibiting network ensembles.

The only motivation is to be able to use an official SuperCollider binary in a school context that doesn't make the firewall complain (and therefore get the IT staff all nervous) :-)

**Comment by telephon** [comment]
yes also in other scenarios it may be a good option.

**Comment by llloret** [comment]
I will have a look at this. I am talking with @bagong about some details to get started. I am also having a look at HID in Windows, so I am not sure how much time I can put into this, but I will try to have it ASAP.

**Comment by llloret** [comment]
Would this be required both for TCP and UDP? I guess so for consistency, but what do you think?

**Comment by llloret** [comment]
I have just issued a PR for this for UDP (#2081). I cannot test it here now, but perhaps someone can checkout my branch and give it a go?

I will try to do the same for TCP later today.

**Comment by llloret** [comment]
The PR supports both udp and tcp.

**Comment by samaaron** [comment]
👍

**Comment by vivid-synth** [comment]
I've spoken to people who have had (maybe valid, maybe not) concerns that running scsynth on their computer might open up security issues -- this would allay those fears (better than the password-with-tcp option)

May I suggest adding this to supernova for consistency?

**Comment by telephon** [comment]
yes, it seems like a simple and good idea.

**Comment by bagong** [comment]
@telephon , could you give an opinion for #2081, the question is whether to merge it to 3.7 or master. The vote is kind of tied atm.

**Comment by telephon** [comment]
ok, done.

**Comment by brianlheim** [comment]
This was fixed by #2095, which was a re-submission of #2081.
